src/apps/replication/lib/replication_failure_detector.cpp

```diff
@@ -62,71 +62,93 @@ replication_failure_detector::~replication_failure_detector(void)
     dsn_group_destroy(_meta_servers.group_handle());
 }
 
-void replication_failure_detector::end_ping(::dsn::error_code err, const fd::beacon_ack& ack, void* context)
+void replication_failure_detector::set_leader_for_test(rpc_address meta)
 {
-    dinfo("end ping result, error[%s], ack.this_node[%s], ack.primary_node[%s], ack.is_master[%s], ack.allowed[%s]",
+    dsn_group_set_leader(_meta_servers.group_handle(), meta.c_addr());
+}
+
+void replication_failure_detector::switch_master_accordingly(dsn::error_code err, const fd::beacon_ack& ack)
+{
+    rpc_address current_master = dsn_group_get_leader(_meta_servers.group_handle());
+    rpc_address next_master;
+    if (err != ERR_OK || ack.primary_node.is_invalid())
+        next_master = dsn_group_forward_leader(_meta_servers.group_handle());
+    else {
+        next_master = ack.primary_node;
+        dsn_group_set_leader(_meta_servers.group_handle(), ack.primary_node.c_addr());
+    }
+
+    if (current_master!=next_master)
+        switch_master(current_master, next_master);
+}
+
+void replication_failure_detector::end_ping(::dsn::error_code err, const fd::beacon_ack& ack, void*)
+{
+    dinfo("end ping result, error[%s], ack.this_node[%s], ack.primary_node[%s], "
+          "ack.is_master[%s], ack.allowed[%s], ack.version[%" PRId64 "]"
+          "ack.time: [%" PRId64 "]",
           err.to_string(), ack.this_node.to_string(), ack.primary_node.to_string(),
-          ack.is_master ? "true" : "false", ack.allowed ? "true" : "false");
+          ack.is_master ? "true" : "false", ack.allowed ? "true" : "false", ack.version,
+          ack.time);
 
-    if (!failure_detector::end_ping_internal(err, ack))
-    {
-        // not applicable
+    zauto_lock l(failure_detector::_lock);
+    dassert(_masters.size()<=1, "currently replica server are not allowed to connect to multiple metas at the same time");
+
+    auto iter = _masters.find(ack.this_node);
+    /* For simplicity, we just ignore all the ack from previous master(s), no matter what they say */
+    if ( iter==_masters.end() ) {
+        dinfo("ping ack from unregistered master, may be staled msg");
+        return;
+    }
+    master_record& cur_record = iter->second;
+    if (cur_record.version != ack.version) {
```

Comment:
this version is not necessary as we already have check on line 108 below. It actually kills certain valid acks that increases unnecessary failure alarms.